sql/src/main/java/org/apache/druid/sql/calcite/external/TableAppendMacro.java

```diff
@@ -0,0 +1,309 @@
+  {
+    @Override
+    public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFailure)
+    {
+      for (int i = 0; i < callBinding.getOperandCount(); i++) {
+        SqlNode operand = callBinding.operand(i);
+        if (!callBinding.isOperandLiteral(i, false)) {
+          if (throwOnFailure) {
+            throw DruidSqlValidator.buildCalciteContextException(
+                "All arguments to APPEND should be literal strings. "
+                    + "Argument #" + i + " is not literal",
+                operand
+            );
+          } else {
+            return false;
+          }
+        }
+
+        SqlTypeName typeName = callBinding.getOperandType(i).getSqlTypeName();
+        if (!SqlTypeFamily.CHARACTER.getTypeNames().contains(typeName)) {
+          if (throwOnFailure) {
+            throw DruidSqlValidator.buildCalciteContextException(
+                "All arguments to APPEND should be literal strings. "
+                    + "Argument #" + i + " is not string",
+                operand
+            );
+          } else {
+            return false;
+          }
+        }
+      }
+      return true;
+    }
+
+    @Override
+    public SqlOperandCountRange getOperandCountRange()
+    {
+      return SqlOperandCountRanges.from(2);
+    }
+
+    @Override
+    public String getAllowedSignatures(SqlOperator op, String opName)
+    {
+      return "APPEND( <TABLE_NAME>, <TABLE_NAME>[, <TABLE_NAME> ...] )";
+    }
+
+    @Override
+    public List<RelDataType> paramTypes(RelDataTypeFactory typeFactory)
+    {
+      RelDataType t = typeFactory.createSqlType(SqlTypeName.VARCHAR);
+      return ImmutableList.<RelDataType>builder().add(t, t).build();
+    }
+
+    @Override
+    public List<String> paramNames()
+    {
+      return ImmutableList.<String>builder().add("tableName", "tableName").build();
+    }
+  }
+
+  @Override
+  public List<String> getParamNames()
+  {
+    return ImmutableList.<String>builder().add("tableName", "tableName").build();
+  }
+
+  @Override
+  public TranslatableTable getTable(SqlOperatorBinding callBinding)
+  {
+    SqlValidator validator = ((SqlCallBinding) callBinding).getValidator();
+    List<RelOptTable> tables = getTables(callBinding, validator.getCatalogReader());
+    AppendDatasourceMetadata metadata = buildUnionDataSource(tables);
+    return new DatasourceTable(
+        metadata.values,
+        metadata,
+        EffectiveMetadata.of(metadata.values)
+    );
+  }
+
+  private List<RelOptTable> getTables(SqlOperatorBinding callBinding, SqlValidatorCatalogReader catalogReader)
+  {
+    List<RelOptTable> tables = new ArrayList<>();
+    for (int i = 0; i < callBinding.getOperandCount(); i++) {
+      if (!callBinding.isOperandLiteral(i, false)) {
+        throw new IllegalArgumentException(
+            "All arguments of call to macro "
+                + "APPEND should be literal. Actual argument #"
+                + i + " is not literal"
+        );
```

Comment:
Are we certain that this error message will make it back to the user?  I'm not sure what phase we are in here.  If we want this message to get to the user, I wonder about making it an `InvalidSqlInput.exception()` instead.